emission/net/ext_service/geocoder/nominatim.py

```diff
@@ -13,12 +13,12 @@
 
 from emission.core.wrapper.trip_old import Coordinate
 try:
-    NOMINATIM_QUERY_URL_env = os.environ.get("NOMINATIM_QUERY_URL", "")
-    logging.info(f"NOMINATIM_QUERY_URL_env: {NOMINATIM_QUERY_URL_env}")
-    NOMINATIM_QUERY_URL = NOMINATIM_QUERY_URL_env if NOMINATIM_QUERY_URL_env != "" else "http://nominatim.openstreetmap.org"
-
+    OPENSTREETMAP_QUERY_URL_env = os.environ.get("OPENSTREETMAP_QUERY_URL", "")
+    logging.info(f"OPENSTREETMAP_QUERY_URL_env: {OPENSTREETMAP_QUERY_URL_env}")
+    OPENSTREETMAP_QUERY_URL = OPENSTREETMAP_QUERY_URL_env if OPENSTREETMAP_QUERY_URL_env != "" else "http://nominatim.openstreetmap.org"
+    print("Open Street Map URL not configured, defaulting to nominatim:") if OPENSTREETMAP_QUERY_URL == "http://nominatim.openstreetmap.org" else print("Open Street Map URL configured!") 
```

Comment:
again, nominatim is software. "http://nominatim.openstreetmap.org" is the openstreetmap hosted version of nominatim. 

`Open Street Map URL not configured, defaulting to nominatim:") if OPENSTREETMAP_QUERY_URL == "http://nominatim.openstreetmap.org"` is wrong at so many levels...

Reply:
This is what I have written in nominatim.py to set up the URL used in production right now:
```
 GEOFABRIK_QUERY_URL_env = os.environ.get("GEOFABRIK_QUERY_URL", "")
    logging.info(f"GEOFABRIK_QUERY_URL_env: {GEOFABRIK_QUERY_URL_env}")

    NOMINATIM_QUERY_URL = (
        GEOFABRIK_QUERY_URL_env if GEOFABRIK_QUERY_URL_env != "" else "http://nominatim.openstreetmap.org"
    )
    print("Geofabrik Query URL not configured, defaulting to OSM") if "nominatim" in NOMINATIM_QUERY_URL else print("Geofabrik Query URL configured!")
```
So the free version of nominatim from OSM is used if the geofabrik url isn't passed in as an environment variable.

Reply:
This is too convoluted. Also people don't have to use GEOFABRIK - they can pick any other nominatim provider.

`nominatim..py` should only use `NOMINATIM_QUERY_URL`
- In production, the `docker-compose` should set `NOMINATIM_QUERY_URL` to the geofabrik API URL
- In testing, you should set `NOMINATIM_QUERY_URL` by setting the environment variable
https://stackoverflow.com/questions/5971312/how-to-set-environment-variables-in-python

Reply:
@nataliejschultz an environment variable is a **variable**. You can get and set it just like any other variable and use it to pass in a parameter
https://github.com/e-mission/e-mission-server/pull/929#discussion_r1341910859

Reply:
I am setting the variable at the top of nominatim.py: 
```
    NOMINATIM_QUERY_URL = os.environ.get("NOMINATIM_QUERY_URL")

    if NOMINATIM_QUERY_URL is None:
        raise Exception("Nominatim query url not configured")
```
I modified both of the functions (make_url_geo and make_url_reverse) in the nominatim.py module to use `NOMINATIM_QUERY_URL` as such:

```
@classmethod
def make_url_geo(cls, address):
    params = {
        "q" : address,
        "format" : "json"
    }
    query_url = NOMINATIM_QUERY_URL + "/search?"
    encoded_params = urllib.parse.urlencode(params)
    url = query_url + encoded_params
    logging.debug("For geocoding, using URL %s" % url)
    return url
```

As of now I am setting the `NOMINATIM_QUERY_URL ` environment variable to the geofabrik url in the docker compose (for testing):
```
environment:
      - DB_HOST=db
      - GFBK_KEY=$GFBK_KEY
      - NOMINATIM_QUERY_URL=https://geofabrik.de/$GFBK_KEY
```

But if we wanted to test the other urls (container/osm urls) with the same functions (ie `eco.make_url_geo` and `make_url_reverse`) we would need to monkey patch the `NOMINATIM_QUERY_URL` variable in the nominatim.py module to the other urls (container and osm). But this also would cause us to have to reset the `eco.NOMINATIM_QUERY_URL` back to the geofabrik url if we then want to compare another `eco` function with 2 different urls. 

I might be greatly overcomplicating this. Do we want the functions in the nominatim.py module to make the same calls with the other urls (container/osm)in the tests at all? Or do you want the tests that compare results between different nominatim services to formulate the call urls outside of nominatim.py, and **only** use the value of `NOMINATIM_QUERY_URL` in nominatim.py? ie this pseudocode:

```
    def test_nominatim_api(self):
        lat, lon = "41.832942092439694", "-71.41558148857203"
        osm_url = OSM_QUERY_URL + "lat=%&lon=%&format=json" % (lat, lon)
        container_url = CONTAINER_URL + "lat=%&lon=%&format=json" % (lat, lon)
        osm_result = requests.get(osm_url)
        container_result = requests.get(container_url)
        geofabrik_result = eco.Geocoder.reverse_geocode(lat, lon)
        self.assertEqual(results)
```
Where the other container URLs are set as variables at the top of TestNominatim.py

Reply:
> I modified both of the functions (make_url_geo and make_url_reverse) in the nominatim.py module to use NOMINATIM_QUERY_URL as such:

You are setting the **python** variable by reading the **environment** variable. This is correct.

> As of now I am setting the NOMINATIM_QUERY_URL environment variable to the geofabrik url in the docker compose (for testing):

This is wrong. You should set two separate environment variables - for the `GEOFABRIK` and the `OSM` URLs.

> we would need to monkey patch the NOMINATIM_QUERY_URL variable in the nominatim.py module to the other urls (container and osm). 

not quite. You would set (no need to monkey patch) the `NOMINATIM_QUERY_URL` environment variable in `TestNominatim.py` and reload the module (`importlib.reload`) before making calls to `eco.Geocoder`
https://github.com/e-mission/e-mission-server/pull/929#discussion_r1346642913

> But this also would cause us to have to reset the eco.NOMINATIM_QUERY_URL back to the geofabrik url if we then want to compare another eco function with 2 different urls.

You would still need to set the `NOMINATIM_QUERY_URL` environment variable back to geofabrik, which is trivial once you are passing in the `GEOFABRIK` URL separately.

>  might be greatly overcomplicating this. 

Yes

> Do we want the functions in the nominatim.py module to make the same calls with the other urls (container/osm)in the tests at all?

No. Why would I want that?

> Or do you want the tests that compare results between different nominatim services to formulate the call urls outside of nominatim.py, and only use the value of NOMINATIM_QUERY_URL in nominatim.py? ie this pseudocode:

Largely yes. Formulate the URL in TestNominatim, set the environment variable, reload the module so that it reads the new environment variable, only use `NOMINATIM_QUERY_URL` in `nominatim.py`

Reply:
> Largely yes. Formulate the URL in TestNominatim, set the environment variable, reload the module so that it reads the new environment variable, only use `NOMINATIM_QUERY_URL` in `nominatim.py`

See [this comment ](https://github.com/e-mission/e-mission-docs/issues/937#issuecomment-1754046911)on how I've addressed this!